### PR TITLE
[serve] Prevent ASGI `receive()` from hanging on disconnects

### LINKED ISSUE #44644: [Serve] Requests to replicas fail to get cancelled and they get stuck at max ongoing requests
### What happened + What you expected to happen

There appears to be a race condition, whereby a replica can get stuck processing a request.

This appears to happen when the HTTPProxy cancels a request before or just after the replica starts processing it:
```
[36m(ProxyActor pid=32578, ip=10.154.83.207)[0m INFO 2024-04-06 23:09:37,136 proxy <ip> 53d547e2-f6c0-410e-ba3a-280dfa33a1d9 <url> proxy.py:1025 - Client for request 53d547e2-f6c0-410e-ba3a-280dfa33a1d9 disconnected, cancelling request.
```

The replica produces these logs, and then is stuck from this point on:
```
INFO 2024-04-06 23:09:37,133 <model> 6yp4j7ox 53d547e2-f6c0-410e-ba3a-280dfa33a1d9 <url> replica.py:1062 - Started executing request to method '__call__'.
WARNING 2024-04-06 23:09:37,135 <model> 6yp4j7ox replica.py:502 - Replica at capacity of max_ongoing_requests=1, rejecting request 4478fffa-e7f2-4333-b58a-f3a240c9fc28.
WARNING 2024-04-06 23:09:38,071 <model> replica.py:502 - Replica at capacity of max_ongoing_requests=1, rejecting request f370c98c-3cfc-436b-ba70-d8ec7263c4d9.
```

Over time more and more replicas get stuck, degrading tail latency.

In our testing, a replica can get stuck simply on the `await raw_request.json()` step.

We have been able to alleviate the issue for us by instead using `await asyncio.wait_for(raw_request.json(), timeout=5.0)`.

After making this change, we can see in the logs that we instead at least timeout and fail the request.
### Versions / Dependencies

Ray 2.10

### Reproduction script

Test deployment:
```python
import time
import asyncio
from ray import serve

@serve.deployment(
    ray_actor_options={
        "num_cpus": 1,
    },
    max_concurrent_queries=1,
    autoscaling_config={
        "min_replicas": 2,
        "max_replicas": 2,
    }
)
class TestModel():
    def __init__(self):
        pass

    async def __call__(self, raw_request):
        # Test #1: ignore the request --> cannot get the replica to get stuck after 20+ mins

        # Test #2: await the request like we currently do --> each replica gets stuck after 2-10 mins
        request = await raw_request.json()

        # Test #3 await the request with a timeout --> cannot get the replicas to get stuck after 20+ mins
        request = await asyncio.wait_for(raw_request.json(), timeout=5.0)

        time.sleep(10)
        return {"message": "Hello, World!"}

app = TestModel.bind()
```

Curl command that includes a 1 second timeout to trigger cancellations, which eventually leads to replicas getting stuck:
```bash
seq 1 1000000 | xargs -n1 -P20 curl -s -o /dev/null --max-time 1 \
      -X POST <url> \           
      -H "Content-Type: application/json" \
      -d '{}'
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

See https://github.com/ray-project/ray/issues/44644 for the user-reported issue.

In https://github.com/ray-project/ray/pull/43713, I changed the behavior of the ASGI `receive` interface to not raise an exception but rather silently hang if the request was not present in the proxy. This was done under the assumption that the code calling `receive` would be subsequently cancelled anyways. However, this assumption was faulty for a few reasons:

- The user may ignore the `CancelledError` but still want to fetch `ASGI` messages.
- Cancellation is best-effort. Specifically, because we are currently using the `run_coroutine_threadsafe` / `concurrent.futures.Executor` interface, there is no way to guarantee that either the scheduling task is cancelled or the resulting object ref is cancelled. In rare cases, neither may happen (see https://github.com/python/cpython/issues/105836).

This PR mitigates the issue by returning an appropriate `disconnect` ASGI message when the proxy no longer has the request in scope.

Note that this is not a perfect solution: we may still drop the messages _prior_ to the disconnect (e.g., the body of the HTTP request) if the replica does not fetch them prior to the client disconnecting.

- Note that this is not a regression from the current behavior and should only happen in practice if the client disconnects very quickly after initiating the request.
- As a follow-up, we should rework this codepath to something more robust. For example we could use a push-based model where the proxy instead pushes messages eagerly to the replica (this was originally very difficult to implement but should be possible now).

TODO before merging:

- [x] Manually verify the fix using the reproduction script in the issue.

## Related issue number

Closes https://github.com/ray-project/ray/issues/44644

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_util.py [hunk 1]
```diff
@@ -217,14 +217,27 @@ class ASGIReceiveProxy:
...
         receive_asgi_messages: Callable[[str], Awaitable[bytes]],
     ):
+        self._type = scope["type"]  # Either 'http' or 'websocket'.
         self._queue = asyncio.Queue()
         self._request_id = request_id
         self._receive_asgi_messages = receive_asgi_messages
         self._disconnect_message = None
 
+    def _get_default_disconnect_message(self) -> Message:
+        """Return the appropriate disconnect message based on the connection type."""
```
Comment (GeneDer) [inline python/ray/serve/_private/http_util.py:231]: Nitpick (non-blocker): Would be nice to also link the specs here https://asgi.readthedocs.io/en/latest/specs/www.html#disconnect-receive-event-ws and https://asgi.readthedocs.io/en/latest/specs/www.html#disconnect-receive-event

File: python/ray/serve/tests/unit/test_http_util.py [hunk 2]
```diff
@@ -125,9 +125,15 @@ async def test_raises_exception(
         with pytest.raises(RuntimeError, match="oopsies"):
             await asgi_receive_proxy()
 
-    async def test_does_not_raise_key_error(
+    async def test_return_disconnect_on_key_error(
```
Comment (GeneDer) [inline python/ray/serve/tests/unit/test_http_util.py:128]: Nitpick (non-blocker): Can also test the websocket codepath to ensure there won't be future regressions.

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/tests/unit/test_http_util.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_http_util.py:137]: just added!

File: python/ray/serve/_private/http_util.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/_private/http_util.py]: done

**Review by shrekris-anyscale — APPROVED** [review]

**Pull request merged**